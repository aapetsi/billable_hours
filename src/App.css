.App {
  text-align: center;
}

.jumbotron {
  background-image: url('https://images.unsplash.com/photo-1544377193-33dcf4d68fb5?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=500&q=60');
  background-size: cover;
}

.text {
  color: rgb(17, 73, 87);
  font-weight: bolder;
  letter-spacing: 1px;
}

.csv-input {
  padding: 10px;
  display: block;
  margin: 15px auto;
  border: 1px solid #ccc;
  border-radius: 5px;
}

.inputs {
  margin: 0 auto;
  display: flex;
  justify-content: space-evenly;
  width: 60%;
  align-items: center;
}

.disabled {
  cursor: not-allowed;
}
