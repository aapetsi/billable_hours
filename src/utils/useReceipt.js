import { useState } from 'react'

export const useReceipt = () => {
  const [timeSheet, setTimeSheet] = useState([])
  const [company, setCompany] = useState('')
  const [receipt, setReceipt] = useState([])

  const papaparseOptions = {
    header: true,
    dynamicTyping: true,
    skipEmptyLines: true,
    transformHeader: header => header.toLowerCase().replace(/\W/g, '_')
  }

  const handleFileUpload = (data, filename) => {
    setReceipt([])
    setTimeSheet([...data])
  }

  const handleCompanyChange = e => {
    setCompany(e.target.value)
    setReceipt([])
  }

  const handleGenerateReceipt = () => {
    // filter companies by their name
    let res = []
    const filteredCompanies = timeSheet.filter(item => item.project === company)
    filteredCompanies.forEach(item => {
      // define start and end dates
      const start = new Date(`${item.date} ${item.start_time}`)
      const end = new Date(`${item.date} ${item.end_time}`)
      const hours = (end.getTime() - start.getTime()) / 3.6e6
      res.push({
        employee_id: item.employee_id,
        billable_rate: item.billable_rate,
        project: item.project,
        date: item.date,
        hours: hours.toFixed(2),
        total: hours * item.billable_rate
      })
    })
    setReceipt([...res])
  }

  return {
    timeSheet,
    company,
    receipt,
    handleGenerateReceipt,
    handleCompanyChange,
    handleFileUpload,
    papaparseOptions
  }
}
