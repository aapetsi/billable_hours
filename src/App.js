import React, { useState } from 'react'
import CSVReader from 'react-csv-reader'
import './App.css'
import 'bootstrap/dist/css/bootstrap.min.css'
import Receipt from './components/Receipt'
import { Jumbotron, Button } from 'react-bootstrap'

const App = () => {
  const [timeSheet, setTimeSheet] = useState([])
  const [company, setCompany] = useState('')
  const [receipt, setReceipt] = useState([])

  const papaparseOptions = {
    header: true,
    dynamicTyping: true,
    skipEmptyLines: true,
    transformHeader: header => header.toLowerCase().replace(/\W/g, '_')
  }

  const handleFileUpload = (data, filename) => {
    setReceipt([])
    setTimeSheet([...data])
  }

  const handleCompanyChange = e => {
    setCompany(e.target.value)
  }

  const handleGenerateReceipt = () => {
    if (!company) {
      alert('Select a company from the drop down list')
    }
    // filter companies by their name
    let res = []
    const filteredCompanies = timeSheet.filter(item => item.project === company)
    filteredCompanies.forEach(item => {
      // define start and end dates
      const start = new Date(`${item.date} ${item.start_time}`)
      const end = new Date(`${item.date} ${item.end_time}`)
      const hours = (end.getTime() - start.getTime()) / 3.6e6
      res.push({
        employee_id: item.employee_id,
        billable_rate: item.billable_rate,
        project: item.project,
        date: item.date,
        hours: hours.toFixed(2),
        total: hours * item.billable_rate
      })
    })
    setReceipt([...res])
  }

  const distinctCompanies = [...new Set(timeSheet.map(x => x.project))]
  const selectCompany = (
    <select onChange={handleCompanyChange} disabled={timeSheet.length === 0}>
      <option value=''>select company</option>
      {distinctCompanies.map((item, idx) => (
        <option key={idx} value={item}>
          {item}
        </option>
      ))}
    </select>
  )
  return (
    <div className='App' data-test='component-app'>
      <Jumbotron className='jumbotron'>
        <h1 className='text'>Welcome to Kratos Billing Systems</h1>
        <p className='text'>
          To begin, upload your time sheet data in CSV format
        </p>
      </Jumbotron>

      <CSVReader
        cssClass='react-csv-input'
        parserOptions={papaparseOptions}
        onFileLoaded={handleFileUpload}
      />
      <Button onClick={handleGenerateReceipt} disabled={timeSheet.length === 0}>
        Generate Receipt
      </Button>

      {selectCompany}
      <br />

      {receipt.length === 0 ? (
        <p>Your receipts will show here</p>
      ) : (
        <Receipt company={company} receipt={receipt} />
      )}
    </div>
  )
}

export default App
