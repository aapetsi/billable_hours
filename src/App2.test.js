import React from 'react'
import ReactDOM from 'react-dom'
import App from './App'
import * as rtl from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import { setHookState } from './utils/setHookState'

afterEach(rtl.cleanup)

let wrapper, csvReader

beforeEach(() => {
  wrapper = rtl.render(<App />)
})

const timeSheet = [
  {
    employee_id: 1,
    billable_rate: 300,
    project: 'google',
    date: '2020-01-20',
    start_time: '08:00',
    end_time: '12:00'
  },
  {
    employee_id: 2,
    billable_rate: 100,
    project: 'facebook',
    date: '2020-01-20',
    start_time: '09:30',
    end_time: '16:50'
  },
  {
    employee_id: 3,
    billable_rate: 250,
    project: 'twitter',
    date: '2020-01-20',
    start_time: '12:00',
    end_time: '16:00'
  },
  {
    employee_id: 1,
    billable_rate: 300,
    project: 'instagram',
    date: '2020-01-21',
    start_time: '10:30',
    end_time: '15:30'
  },
  {
    employee_id: 2,
    billable_rate: 100,
    project: 'twitter',
    date: '2020-01-21',
    start_time: '08:00',
    end_time: '14:15'
  },
  {
    employee_id: 3,
    billable_rate: 250,
    project: 'yahoo',
    date: '2020-01-21',
    start_time: '13:00',
    end_time: '15:00'
  },
  {
    employee_id: 1,
    billable_rate: 300,
    project: 'uber',
    date: '2020-01-22',
    start_time: '09:00',
    end_time: '11:00'
  },
  {
    employee_id: 2,
    billable_rate: 100,
    project: 'slydepay',
    date: '2020-01-22',
    start_time: '08:00',
    end_time: '17:00'
  },
  {
    employee_id: 3,
    billable_rate: 250,
    project: 'linkedin',
    date: '2020-01-22',
    start_time: '12:00',
    end_time: '16:00'
  },
  {
    employee_id: 1,
    billable_rate: 300,
    project: 'linkedin',
    date: '2020-01-23',
    start_time: '11:00',
    end_time: '15:00'
  },
  {
    employee_id: 2,
    billable_rate: 100,
    project: 'instagram',
    date: '2020-01-23',
    start_time: '13:00',
    end_time: '16:00'
  },
  {
    employee_id: 3,
    billable_rate: 250,
    project: 'facebook',
    date: '2020-01-23',
    start_time: '15:00',
    end_time: '17:00'
  },
  {
    employee_id: 1,
    billable_rate: 300,
    project: 'slydepay',
    date: '2020-01-24',
    start_time: '09:00',
    end_time: '16:00'
  },
  {
    employee_id: 2,
    billable_rate: 100,
    project: 'whatsapp',
    date: '2020-01-24',
    start_time: '12:00',
    end_time: '15:00'
  },
  {
    employee_id: 3,
    billable_rate: 250,
    project: 'whatsapp',
    date: '2020-01-24',
    start_time: '11:00',
    end_time: '16:00'
  }
]
const company = 'facebook'
const receipt = [
  {
    employee_id: 2,
    billable_rate: 100,
    project: 'facebook',
    date: '2020-01-20',
    hours: '7.33',
    total: 733.3333333333333
  },
  {
    employee_id: 3,
    billable_rate: 250,
    project: 'facebook',
    date: '2020-01-23',
    hours: '2.00',
    total: 500
  }
]
const show = false

describe('App component', () => {
  test('should render without crashing', () => {
    const div = document.createElement('div')
    ReactDOM.render(<App />, div)
  })

  test('should match the snapshot', () => {
    expect(wrapper.container).toMatchSnapshot()
  })

  test('should render homepage heading', () => {
    expect(
      wrapper.queryByText(/welcome to kratos billing systems/i)
    ).toBeInTheDocument()
  })

  test('button should be disabled', () => {
    const receiptBtn = wrapper.queryByText(/generate receipt/i)
    expect(receiptBtn.disabled).toBe(true)
  })

  test('should render the company drop down', () => {
    const select = rtl.getByTestId(wrapper.container, 'select')
    console.log(select.textContent)
    expect(select).toHaveTextContent(/select company/i)
  })
})
