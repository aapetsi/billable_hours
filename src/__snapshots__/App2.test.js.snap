// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should match the snapshot 1`] = `
<div>
  <div
    class="App"
    data-test="component-app"
  >
    <div
      class="jumbotron jumbotron"
    >
      <h1
        class="text"
      >
        Welcome to Kratos Billing Systems
      </h1>
      <p
        class="text"
      >
        To begin, upload your time sheet data in CSV format
      </p>
    </div>
    <div
      class="react-csv-input"
    >
      <input
        accept=".csv, text/csv"
        class="csv-input"
        type="file"
      />
    </div>
    <div
      class="container"
    >
      <div
        class="inputs"
      >
        <div>
          <button
            class="btn btn-primary"
            data-test="component-generate-receipt"
            disabled=""
            type="button"
          >
            Generate Receipt
          </button>
        </div>
        <div
          data-testid="select"
        >
          <select
            disabled=""
          >
            <option
              value=""
            >
              Select Company
            </option>
          </select>
        </div>
      </div>
      <br />
      <p
        data-test="component-receipt"
        data-testid="receipt-none"
      >
        Your receipts will show here
      </p>
    </div>
  </div>
</div>
`;
